### ISSUE TITLE
TTY Error running interactive docker on Bash on Windows

### ISSUE THREAD

**Issue by waleedka** [docker/for-win#1588]
A quick search online showed that a lot of people are frustrated about this issue. Basically, I'm trying to run Docker for Windows from the Bash on WSL (Windows Subsystem for Linux) on Windows 10 Creator Edition.

```
 docker.exe run -it hello-world
```
Returns this error:
```
the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'
```

The suggested solutions typically require installing external tools to work around the issue. I prefer not to prefix all my docker commands with `winpty` because then my scripts won't run on Linux boxes. And, actually, I prefer not to have to install `winpty` or any other tool just to work-around this issue. Can someone at Docker fix the issue at the root please? 

Based on [this thread](https://github.com/Microsoft/WSL/issues/2379), it seems this is a Docker issue. Quoting:
> This is a docker.exe issue, I believe. It is not aware of the new console functionality and probably duplicates it to some degree. We saw similar problems on Win32-OpenSSH. Basically, when compiling docker.exe, they need to flip a few console flags up on the IO handles and disable their built-in terminal emulation code to allow the Windows console to take over processing.

Please look into it. You'd save your community a lot of frustration.

**Comment by AnisAli** [comment]
Run without .exe
`
docker  run -it hello-world
`

**Comment by jtrutwin** [comment]
Um, No different running without the .exe, not sure why you expected that would work


jtrutwin@DESKTOP-PAK1BAG MINGW64 ~/git-projects/Hive (master)
$ docker run -it hello-world
the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'


Also running into this issue, hoping there's a fix down the road.  Thanks.

**Comment by whfamily2006** [comment]
Anybody find the solution to this issue? I am facing this one too.

**Comment by samoldenburg** [comment]
Any update on this issue?

**Comment by ray1007** [comment]
I had the same problem. Below are my findings: 

The command ```docker.exe run -it hello-world``` also failed in my WSL bash.
However I found the same command worked in both cmd and Powershell. (See below)
![image](https://user-images.githubusercontent.com/7671322/42927696-d723e484-8b67-11e8-9c49-d8a35a95dc5a.png)
![image](https://user-images.githubusercontent.com/7671322/42928063-e958d492-8b68-11e8-8bc2-938ad496c7e1.png)
![image](https://user-images.githubusercontent.com/7671322/42927651-b396caf4-8b67-11e8-843a-b22daae21fe5.png)

Perhaps one should not use ```docker.exe``` in WSL. I followed this [guide](https://medium.com/@sebagomez/installing-the-docker-client-on-ubuntus-windows-subsystem-for-linux-612b392a44c4) to install docker client on WSL. And the command ```docker run -it hello-world``` executed with no error.
![image](https://user-images.githubusercontent.com/7671322/42928629-6ee82e2c-8b6a-11e8-8850-122e1b8114e3.png)

Still I haven't figured out why ```docker.exe``` works only on windows shells. I'd appreciate if someone could explain it! :)

Tzu-Ray

**Comment by tomasaschan** [comment]
/remove-lifecycle stale

I'm also wondering about this. Would be nice if this worked seamlessly in WSL!

**Comment by mrbarua** [comment]
On git bash just type:
alias docker="winpty docker"
and tty will work

after starting the docker I mean when it will run 
do a docker login from powershell and then use git-bash

then create docker conatainer
docker container run -it -p 80:80 nginx

**Comment by nngo** [comment]
If your application does not need to send output text/etc to another application (e.g. for piping), then you don't need to run using the `--tty` or `-t` option to create a pseudo-terminal (i.e. for **STDOUT**). Additionally, if the application does not need to read from **STDIN** (then you can also omit the `--interactive` or `-i` option) e.g.

```shell
# this should still output to your console
docker run hello-world
```

In response to https://github.com/docker/for-win/issues/1588#issuecomment-449421977
When running the nginx web server, most of the time you don't need the `-it` (aka **STDIIN** via `--interactive` and **STDOUT** via `--tty` for the psuedo-tty options)
```shell
docker run -p 80:80 nginx
```

**Comment by FDiskas** [comment]
I tried like so. Worked :joy_cat:
```winpty docker run hello-world```

**Comment by judgej** [comment]
April 2019 - still seems to be a problem on Windows 10. `docker` cannot be run interactively without running it through `winpty`.

**Comment by Ruegen** [comment]
same error here - I have to run it though winpty

**Comment by JoanComasFdz** [comment]
Just run into this error using zsh in Unbunty in Hyper:
- Windows 10 1709 Build 16299.1087
- Docker Desktop Community 2.0.0.3 (31259) Build 8858db3 Engine 18.09.2.

When I run the command in cmd.exe it works.

**Comment by siochs** [comment]
Just ommitting `-it` because it's not required by the container is not a solution:
```shell
$ docker run --rm -it debian:latest /bin/bash
the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'
```
Using `winpty` wrapper is not acceptable from my point of view, because it messes up with parsing the command line and with the environment variables:
```shell
$ winpty docker run --rm -it debian:latest /bin/bash
C:/Program Files/Docker/Docker/Resources/bin/docker.exe: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "exec: \"C:/Users/foo/AppData/Local/Programs/Git/usr/bin/bash.exe\": stat C:/Users/foo/AppData/Local/
Programs/Git/usr/bin/bash.exe: no such file or directory": unknown.
```

**Comment by LeeRuns** [comment]
Still does not work for me.  

_**here is what i need to run from git bash on windows**_
`~/$ winpty docker -it run simcompdocker /bin/bash
stdin is not a tty`

_**tried to drop the -i bc it is griping about input so i changed to this**_
`~/$ winpty docker -t run simcompdocker /bin/bash
stdin is not a tty`

_**tried to drop all stdin stdout related work**_
`~$ winpty docker run hello-world
stdin is not a tty
`
**setup is**
1. git-bash
2. docker full windows install

**Comment by movielovers** [comment]
> Still does not work for me.
> 
> _**here is what i need to run from git bash on windows**_
> `~/$ winpty docker -it run simcompdocker /bin/bash stdin is not a tty`
> 
> _**tried to drop the -i bc it is griping about input so i changed to this**_
> `~/$ winpty docker -t run simcompdocker /bin/bash stdin is not a tty`
> 
> _**tried to drop all stdin stdout related work**_
> `~$ winpty docker run hello-world stdin is not a tty `
> **setup is**
> 
> 1. git-bash
> 2. docker full windows install

Hello I just dropped the full path the bash and it worked for me
winpty docker -it run simcompdocker bash

> Just ommitting `-it` because it's not required by the container is not a solution:
> 
> ```shell
> $ docker run --rm -it debian:latest /bin/bash
> the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'
> ```
> 
> Using `winpty` wrapper is not acceptable from my point of view, because it messes up with parsing the command line and with the environment variables:
> 
> ```shell
> $ winpty docker run --rm -it debian:latest /bin/bash
> C:/Program Files/Docker/Docker/Resources/bin/docker.exe: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "exec: \"C:/Users/foo/AppData/Local/Programs/Git/usr/bin/bash.exe\": stat C:/Users/foo/AppData/Local/
> Programs/Git/usr/bin/bash.exe: no such file or directory": unknown.
> ```

I didn't put the full path in for bash and it worked for me
 winpty docker run --rm -it debian:latest bash

**Comment by LeeRuns** [comment]
bump

**Comment by sbkg0002** [comment]
Bump

**Comment by siochs** [comment]
> I didn't put the full path in for bash and it worked for me
> winpty docker run --rm -it debian:latest bash

Okay, but this appears to me more like a workaround than a solution. Based on your hint, I found the following commands also working (which for the moment confuses me):
` winpty docker run --rm -it debian:latest bin/bash`
` winpty docker run --rm -it debian:latest //bin//bash`
`winpty docker run --rm -it debian:latest //bin/bash`
Also things like
`winpty docker run --rm -it debian:latest //bin/bash -c "echo 'hello'"`
work...

I don't know. Using `winpty` feels for me like installing a `keyboard-driver-for-notepad`.

**Comment by movielovers** [comment]
> Still does not work for me.
> 
> _**here is what i need to run from git bash on windows**_
> `~/$ winpty docker -it run simcompdocker /bin/bash stdin is not a tty`
> 
> _**tried to drop the -i bc it is griping about input so i changed to this**_
> `~/$ winpty docker -t run simcompdocker /bin/bash stdin is not a tty`
> 
> _**tried to drop all stdin stdout related work**_
> `~$ winpty docker run hello-world stdin is not a tty `
> **setup is**
> 
> 1. git-bash
> 2. docker full windows install
----------------------------------------------------------
winpty docker -it run simcompdocker /bin/bash

this is incorrectso the -it goes after the run so 
winpty docker run -it simcompdocker bash

Either way if even easy commands are failing like hello-world.  Almost looks like something is not installed right.

$ winpty docker run hello-world
Hello from Docker!

$ docker run hello-world
Hello from Docker!
This message shows that your installation appears to be working correctly.

didnt even need winpty to run hello-world
I would say if its feasible maybe reinstall if you dont lose too much

**Comment by dsebastien** [comment]
This is the alias I'm using, which seem to do the trick: `docker(){(export MSYS_NO_PATHCONV=1; "winpty" "docker" "$@")}`

**Comment by osoese** [comment]
#!/bin/bash
docker run --name myshellbox --privileged -d shellbox
TEST_IT=$(docker exec myshellbox echo "Hello from container!")
echo ${TEST_IT}

this two step approach to get the output is what I use as a workaround when scripting to container

**Comment by muthuTheerthan** [comment]
Is there any fix ?

issue:
--------
Docker version:  19.03.1 (windows)

Invoking container from CMD: (works fine)

C:\Users\muthu>docker exec -it c31d1f693b74 /bin/bash
root@c31d1f693b74:/#

Same thing from git bash# (Not working)

$ docker exec -it c31d1f693b74 /bin/bash
the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'

60135734@EHL5CG72513QQ MINGW64 ~ (master)
$

**Comment by vagnerx** [comment]
> I tried like so. Worked 😹
> `winpty docker run hello-world`

Worked fine for me too

**Comment by wave-rider** [comment]
I am running latest Windows 10 Education
docker.exe run -it hello-world 
is working OK
![image](https://user-images.githubusercontent.com/25537219/68074780-e4e3c480-fe03-11e9-81d0-108ee5d89646.png)

**Comment by aciddude** [comment]
@ray1007 

.exe programs are usually windows only.  So the reason why  your docker.exe only runs in windows shells is because it's for windows 

WSL or the windows subsystem linux is kind of like a fakeVM that runs a shitty customised version of ubuntu 16.04 

So if you're using WSL, you need to install docker the linux way and then just ruck `docker` and not `docker.exe`

**Comment by siochs** [comment]
> So if you're using WSL, you need to install docker the linux way and then just ruck `docker` and not `docker.exe`

Actually using WSL was the only **solution** that worked for me. You need docker for windows installed and the service running. Make the service available on port 2375. Next you need in your ubuntu WSL setup a docker installation, but make sure not to start the docker daemon in there. Place the following line into ~/.bashrc: `export DOCKER_HOST=tcp://localhost:2375`. Now you can operate as usual: `docker run --rm -it debian:latest /bin/bash`. Keep in mind the container will be handled by the Windows host, not the WSL. This also affects routing. You need to add a static route when you try to reach a running container by its local IP.

> I tried like so. Worked 😹
> `winpty docker run hello-world`

So? Have you tried with `-it`? Have you tried to get a bash inside a container? Running hello world is not the issue here, it is about interactive terminals and winpty.

**Comment by aprilmintacpineda** [comment]
I encountered this when I was trying to run ubuntu with `docker run -ti ubuntu /bin/bash`. It said `the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'`. I was on my windows machine using git-bash. When I tried the same command on windows CMD, it worked. Any insights on this?

**Comment by TavenYin** [comment]
This problem occurs when I use git bash, but when I switch to cmd, there is no problem

**Comment by omar-bakhsh** [comment]
Git Alias Overview 
https://www.atlassian.com/git/tutorials/git-alias

**Comment by eric-orenge** [comment]
This works 

docker exec -it [container_name] bash

**Comment by giusepe22** [comment]
winpty é para acesso bash do windows

**Comment by cosminbosutar** [comment]
Have a look [here](https://github.com/zeit/hyper/issues/2888).
You only need to add "winpty" in start of your cmd.
Ex: `$ winpty docker.exe run -it --rm ubuntu:14.04 /bin/bash`

**Comment by iamkoch** [comment]
I've come by this thread with the same problem as the initial poster.

I am using windows 10, but want to write a script that will work on both windows _and_ linux.

Prepending `winpty` is not a good enough solution, so I have come up with this:

```
function runDocker() {
    if command -v winpty2 &> /dev/null; then
        winpty "$@"
    else
        "$@"
    fi
}

runDocker docker run --privileged -it ...
```

Has anyone done the same and had any issues with it?

**Comment by lbogdan** [comment]
The solution is simpler (and weirder?) than you might think: `Git Bash` by default runs `C:\Program Files\Git\git-bash.exe`:

![image](https://user-images.githubusercontent.com/2083930/72683631-bcee2e00-3ae1-11ea-870e-f9aec05dd930.png)

which indeed errors when you run a `docker` command that requires a tty:

![image](https://user-images.githubusercontent.com/2083930/72683683-09396e00-3ae2-11ea-99d3-26160bd42f68.png)

BUT, if you instead run `C:\Program Files\Git\bin\bash.exe`:

![image](https://user-images.githubusercontent.com/2083930/72683708-47cf2880-3ae2-11ea-9862-6a5cab9d3e9e.png)

then it just works:

![image](https://user-images.githubusercontent.com/2083930/72683730-7b11b780-3ae2-11ea-9373-21b513a54bee.png)

**Comment by LeeRuns** [comment]
hmm, but that still limits user to the git bash interface, many people use VS code, emacs, etc. which then use the git-bash.exe

Is there some kind of env variable that is either being loaded or not being loaded that allows this to work?

**Comment by lbogdan** [comment]
Actually, it works out-of-the box in VSCode:

![image](https://user-images.githubusercontent.com/2083930/72750450-8b4e9300-3bc5-11ea-9936-974e26bf2836.png)

after configuring it to use `Git Bash` as the terminal shell:

![image](https://user-images.githubusercontent.com/2083930/72750612-03b55400-3bc6-11ea-8db0-a12c9104ef1a.png)

You can see that VSCode also uses `C:\Program Files\Git\bin\bash.exe` in that list.

**Comment by LeeRuns** [comment]
but `C:\Program Files\Git\bin\bash.exe` is what emacs is pointing to on my machine.  Still throws the winpty error. In your screenshots above you had `C:\Program Files\Git\bin\bash.exe --cd-to-home`

Is that fixing or was it causing the problem on your machine?

maybe i could do something like
`C:Windows\System32\cmd.exe C:\Program Files\Git\bin\bash.exe`
to make cmd prompt execute git-bash, which would replicate calling from run?

**Comment by lbogdan** [comment]
> In your screenshots above you had C:\Program Files\Git\bin\bash.exe --cd-to-home

Which screenshot are you referring to? The only one having `--cd-to-home` is `"C:\Program Files\Git\git-bash.exe" --cd-to-home`, not `...\Git\bin\bash.exe`.

**Comment by colthreepv** [comment]
as written by @siochs before:
> Just ommitting `-it` because it's not required by the container is not a solution:
> 
> ```shell
> $ docker run --rm -it debian:latest /bin/bash
> the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'
> ```
> 
> Using `winpty` wrapper is not acceptable from my point of view, because it messes up with parsing the command line and with the environment variables:
> 
> ```shell
> $ winpty docker run --rm -it debian:latest /bin/bash
> C:/Program Files/Docker/Docker/Resources/bin/docker.exe: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "exec: \"C:/Users/foo/AppData/Local/Programs/Git/usr/bin/bash.exe\": stat C:/Users/foo/AppData/Local/
> Programs/Git/usr/bin/bash.exe: no such file or directory": unknown.
> ```

the best workaround I could find is using a *better* winpty.
TLDR; [download release from here ](https://github.com/rprichard/winpty/releases) and put it in a priority folder, ex: `~/bin/winpty.exe`  
```shell
$ where winpty
C:\Users\valerio\bin\winpty.exe
C:\Program Files\Git\usr\bin\winpty.exe
```

at this point the escaping will be correct ✔️
```shell
$ winpty echo "/data"
/data
```

instead of the dreaded 🤬
```shell
$ winpty echo "/data"
C:/Program Files/Git/data
```

For a longer explanation, refer to [this issue on MSYS2-packages](https://github.com/msys2/MSYS2-packages/issues/411), and specifically [this comment](https://github.com/msys2/MSYS2-packages/issues/411#issuecomment-372585320)


This is such a **LONG** standing issue, by now I could apply for a paleontologist job.  
My suggestion would be to approve a PR to document it on [install.md](https://github.com/docker/docker.github.io/blob/master/docker-for-windows/install.md) or [faqs.md](https://github.com/docker/docker.github.io/blob/master/docker-for-windows/faqs.md)

**Comment by universidadbackend** [comment]
This work for me on windows: winpty docker run -it ubuntu bash.

**Comment by gihanmu** [comment]
> alias docker="winpty docker"

alias docker="winpty docker"
worked for me

**Comment by jjqq2013** [comment]
Just share a tip for Git Bash for Windows,

"Git Bash for Windows" ships a GUI terminal (right picture), called Git Bash, it does not provide tty.
Another CUI program is ""C:\Program Files\Git\bin\bash.exe" (see left picture), it works perfect.

![image](https://user-images.githubusercontent.com/5234060/79420156-4c2cb400-7ff3-11ea-99c3-fde722cbeddd.png)

**Comment by JoseAburt** [comment]
@jjqq2013 the output says you to put the prefix winpty: winpty docker exec -it ubuntu:<version> bash

**Comment by jjqq2013** [comment]
@JoseAburt  thank you for your kind reply.  I just find that directly running "C:\Program Files\Git\bin\bash.exe" will have no such problem.

**Comment by grimimirg** [comment]
If this redundancy could be of any help:

As already somebody says above, inside `.bashrc` I used to write `alias docker="winpty docker"` like this:

![image](https://user-images.githubusercontent.com/15869076/81103305-223f2100-8f11-11ea-8901-6b9f2f470bd2.png)

then close and open git bash, and then:

![image](https://user-images.githubusercontent.com/15869076/81103543-7813c900-8f11-11ea-9a67-a95eb6fec353.png)

**Comment by kendag** [comment]
I had this issue and fixed it by updating minty through the Cygwin installer:

mintty 3.1.4 (x86_64-pc-cygwin)

**Comment by inspiretk** [comment]
This worked for me:
In git bash windows 10, bind the word docker to "winpty docker"
alias docker="winpty docker"

Then ssh into a container like so:
docker exec -it containerName bash

I was using docker exec nginx-test /bin/bash and this doesnt work. Just remove the /bin/ and just use bash.

Summary: In git bash on windows 10. Run these commands:
alias docker="winpty docker"
docker exec -it containerName bash

Thanks everyone for helping!

**Comment by lbogdan** [comment]
> I was using `docker exec nginx-test /bin/bash` and this doesnt work.

@inspiretk That is because of MinGW's (which Git Bash uses) [Posix path conversion](http://www.mingw.org/wiki/Posix_path_conversion). While only using `bash` works here (because `/bin` is in `$PATH`), if you really need an absolute path you can do either:

```sh
$ docker exec -it containerName //bin/bash
```

(note the double `//`), or

```sh
$ MSYS_NO_PATHCONV=1 docker exec -it containerName /bin/bash
```

**Comment by LeeRuns** [comment]
i switched to new newest version of msys2 and started using mingw based on that (instead of the one that comes with git) and I am having much better success.  Just an fyi for everyone interested in different method

**Comment by robertwt7** [comment]
This occurs on WSL as well...

`alias aws='winpty docker run --rm -it amazon/aws-cli'`

gave me samer error when running aws cli. I don't want to use git-bash :(

**Comment by Dhruvsahu9** [comment]
winpty docker run -it ubuntu bash

**Comment by Dhruvsahu9** [comment]
this comand worked for me,try using winpty

**Comment by bugslifesolutions** [comment]
Found a work-around (based on above winpty suggestion).

### Scenario:
Run a bash script containing TTY (input/output) commands using mingw64/git bash.

### Successful Command
`winpty bash .//bin//package-link api-plugin-inventory-simple`

Result:
```
\nUsing local package path /c/data/gitrepo2/reaction-development-platform/api-plugins/api-plugin-inventory-simple
If this is not correct, specify the correct path as the second argument.\n
Making directory /home/node/copied-packages/api-plugin-inventory-simple in 'api' container
Copying local api-plugin-inventory-simple package code into container path /home/node/copied-packages/api-plugin-inventory-simple ...
Linking package into API...
```

### Unsuccessful Command (script as command)
`$ winpty .//bin//package-link api-plugin-inventory-simple`

Result:

```bash
winpty: error: cannot start 'C:/data/gitrepo2/reaction-development-platform/reaction/bin/package-link api-plugin-inventory-simple': %1 is not a valid Win32 application. (error 0xc1)
```

### Unsuccessful Command (no winpty, directly invoke bash script)
`$ ./bin/package-link api-plugin-inventory-simple`

Result:
```bash
\nUsing local package path /c/data/gitrepo2/reaction-development-platform/api-plugins/api-plugin-inventory-simple
If this is not correct, specify the correct path as the second argument.\n
Making directory /home/node/copied-packages/api-plugin-inventory-simple in 'api' container
the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'
```
Root cause of TTY error:
`docker-compose exec api sh -c "mkdir -p ${container_destination_path}"`

### Referenced Bash Script
```bash
#!/usr/bin/env bash

# Please Use Google Shell Style: https://google.github.io/styleguide/shell.xml

# ---- Start unofficial bash strict mode boilerplate
# http://redsymbol.net/articles/unofficial-bash-strict-mode/
set -o errexit  # always exit on error
set -o errtrace # trap errors in functions as well
set -o pipefail # don't ignore exit codes when piping output
set -o posix    # more strict failures in subshells
# set -x          # enable debugging

IFS="$(printf "\n\t")"
# ---- End unofficial bash strict mode boilerplate

package_name=$1
package_path=$2
container_id="$(docker-compose ps -q api)"
container_copied_packages_path="/home/node/copied-packages"
container_destination_path="${container_copied_packages_path}/${package_name}"

if [[ -z "${package_path}" ]]; then
  package_name_without_org="${package_name/#@reactioncommerce\/}"
  package_path="../api-plugins/${package_name_without_org}"
  full_package_path="$(cd ${package_path} && pwd)"
  echo "\nUsing local package path ${full_package_path}"
  echo "If this is not correct, specify the correct path as the second argument.\n"
fi

echo "Making directory ${container_destination_path} in 'api' container"
docker-compose exec api sh -c "mkdir -p ${container_destination_path}"
echo "Copying local ${package_name} package code into container path ${container_destination_path} ..."
docker cp "${package_path}/." "${container_id}:${container_destination_path}"

# Then npm link into this project
echo "Linking package into API..."
docker-compose exec api sh -c "npm config set prefix /home/node/npm && cd /usr/local/src/app && npm link "${container_destination_path}""

# Fool nodemon into thinking something has changed so that it restarts.
# Touch first file found in /src with .js extension
echo "Restarting API..."
docker-compose exec api sh -c "touch -c $(ls ./src/*.js | head -n1)"
```

**Comment by bugslifesolutions** [comment]
> @JoseAburt thank you for your kind reply. I just find that directly running "C:\Program Files\Git\bin\bash.exe" will have no such problem.

**I wish there was a way to highlight the simplest answer... @JoseAburt  you nailed it!**  
Thank you for sharing.  Works like a charm.


## Simple Solution
run \<git for windows root\>\bin\bash.exe

### Successful Command
`$ ./bin/package-link api-plugin-inventory-simple`

### Result
```bash
Using local package path /c/data/gitrepo2/reaction-development-platform/api-plugins/api-plugin-inventory-simple
If this is not correct, specify the correct path as the second argument.\n
Making directory /home/node/copied-packages/api-plugin-inventory-simple in 'api' container
Copying local api-plugin-inventory-simple package code into container path /home/node/copied-packages/api-plugin-inventory-simple ...
```

### The process tree

![image](https://user-images.githubusercontent.com/15253907/91645971-6bfbcc00-ea18-11ea-89e2-e9670ae31a3d.png)

**Comment by baerrach** [comment]
Combining a few answers as the docker tutorial commands:

```
docker exec e60043f6fcd3 cat /data.txt
docker run -it ubuntu ls /
```

Fail with the linux to windows name mangling and the tty emulation in git for windows bash.

```
# https://github.com/docker/toolbox/issues/673#issuecomment-355275054
# Workaround for Docker for Windows in Git Bash.
docker()
{
        (export MSYS_NO_PATHCONV=1; winpty "docker.exe" "$@")
}

# Workaround for tty issues and docker
# See:
# * https://github.com/msys2/MSYS2-packages/issues/411#issuecomment-372585320
# * https://github.com/docker/for-win/issues/1588 TTY Error running interactive docker on Bash on Windows
# * https://github.com/rprichard/winpty/issues/125 winpty not respecting non-path-munging env vars, or aliases
# Install newer version of WINPTY as per https://github.com/docker/for-win/issues/1588#issuecomment-594938988
export WINPTY_HOME=/C/winpty-0.4.3-msys2-2.7.0-x64
export PATH=${WINPTY_HOME}/bin:$PATH
```

**Comment by Dhruvsahu9** [comment]
thanks

**Comment by georgettica** [comment]
This is still an issue, is the fix on the roadmap?

**Comment by shyaboi** [comment]
I have experienced this error, similar to @ray1007 
![2021-02-02_19h21_18](https://user-images.githubusercontent.com/59028381/106688703-d4347800-658b-11eb-8b9b-5ad4b8eb551e.png)
Just like @ray1007 above, I just tried `docker run -it busybox sh`
and got this error
`the input device is not a TTY.  If you are using mintty, try prefixing the command with 'winpty'`
and as well, it works in my CMD and powershell fine, as you can see the containers running.
